dhis2/dhis2-core
# fix: MinIO - Retrieving and updating an App work correctly [DHIS2-17197]

# Issues when using MinIO for storage
1. Endpoint: `GET` `/dhis-web-{app}/index.html`
    Retrieving an `App` to display in the UI results in a blank page (`200` response with `Content-Length=0`)

2. Endpoint: `POST` `api/appHub/{appId}`
    When updating an `App` to a different version, the process failed with errors:
```text
Failed to install app: error deleting directory apps/capture_100.70.9// in containerName: dhis2
Object name contains unsupported characters
```

# Causes
1. `Content-Length=0` when resource is of type `URL` (`File` works fine)
3. extra `/` in path. The blob storage does not need to call `JCloudStore#deleteDirectory`. The code to remove a blob is already there to use.

# Fixes
1. Resolve the `Content-Length` when the resource is a `URL`  (`File` resource behaviour is unchanged)
2. Branched the existing code for deleting directories/blobs which was used for all storage types, into
    a. File (deletes the filesystem directory & its contents)
    b. Blob (removes the blob, no longer calls `JCloudStore#deleteDirectory` which was the origin of the error) 

# Testing
## Automated
Added MinIO integration tests
- install and update `App`
- get `URL` resource content length

Added some extra provider unit tests also.

## Manual
Tested using the following scenario
- Install Capture `App` (`latest`)
- Install Capture `App` (`100.70.6`)
- Verify Capture `App` opens (index.html received with `Content-Length` greater than 0)
- previous Capture `App` version now removed from storage

Tested using MinIO and default filesystem storage.

# Info
UrlConnection uses caching by default.

Add MinIO TestContainer config.
To use just add `@ContextConfiguration(classes = {MinIOConfiguration.class})` to your integration test.

# Next
File system deletion of directory uses a deprecated method (`BlobStore#deleteDirectory`). Not a quick fix, will look to see if we can use something else in a separate PR chore.
